sdk/core/System.ClientModel/src/Convenience/ClientResultException.cs

```diff
@@ -0,0 +1,122 @@
+using System.IO;
+using System.Runtime.Serialization;
+using System.Text;
+
+namespace System.ClientModel;
+
+[Serializable]
+public class ClientResultException : Exception, ISerializable
+{
+    private const string DefaultMessage = "Service request failed.";
+
+    private readonly PipelineResponse? _response;
+    private int _status;
+
+    /// <summary>
+    /// Gets the HTTP status code of the response. Returns. <code>0</code> if response was not received.
+    /// </summary>
+    public int Status
+    {
+        get => _status;
+        protected set => _status = value;
+    }
+
+    public ClientResultException(PipelineResponse response, string? message = default, Exception? innerException = default)
+        : base(GetMessage(response, message), innerException)
+    {
+        ClientUtilities.AssertNotNull(response, nameof(response));
+
+        _response = response;
+        _status = response.Status;
+    }
+
+    public ClientResultException(string message, Exception? innerException = default)
+        : base(message, innerException)
+    {
+        _status = 0;
+    }
+
+    /// <summary>
+    /// TBD
+    /// </summary>
+    /// <param name="info"></param>
+    /// <param name="context"></param>
+    protected ClientResultException(SerializationInfo info, StreamingContext context)
+        : base(info, context)
+    {
+        _status = info.GetInt32(nameof(Status));
+    }
+
+    /// <inheritdoc />
+    public override void GetObjectData(SerializationInfo info, StreamingContext context)
+    {
+        ClientUtilities.AssertNotNull(info, nameof(info));
+
+        info.AddValue(nameof(Status), Status);
+
+        base.GetObjectData(info, context);
+    }
+
+    public PipelineResponse? GetRawResponse() => _response;
+
+    // Create message from response if available, and override message, if available.
+    private static string GetMessage(PipelineResponse response, string? message)
+    {
+        // Setting the message will override extracting it from the response.
+        if (message is not null)
+        {
+            return message;
+        }
+
+        if (!response.TryGetBufferedContent(out _))
+        {
+            BufferResponse(response);
+        }
+
+        StringBuilder messageBuilder = new();
+
+        messageBuilder
+            .AppendLine(DefaultMessage)
+            .Append("Status: ")
+            .Append(response.Status.ToString(CultureInfo.InvariantCulture));
+
+        if (!string.IsNullOrEmpty(response.ReasonPhrase))
+        {
+            messageBuilder.Append(" (")
+                .Append(response.ReasonPhrase)
+                .AppendLine(")");
+        }
+        else
+        {
+            messageBuilder.AppendLine();
+        }
+
+        // Content or headers can be obtained from raw response so are not added here.
+
+        return messageBuilder.ToString();
+    }
+
+    private static void BufferResponse(PipelineResponse response)
+    {
+        if (response.ContentStream is null)
```

Comment:
@KrzysztofCwalina: should this also check if the response is not already buffered?
https://github.com/Azure/azure-sdk-for-net/pull/41009#discussion_r1442251806

Reply:
We do that on line 77, but let me see if I can rework this with the refactor described to move the buffering into PipelineResponse itself.

Reply:
Addressed by moving the response buffering logic onto PipelineResponse.

Reply:
Then can we debug assert here that it's not already buffered?